src/Plugin/Validators/ValidHeaders.php

```diff
@@ -86,18 +128,151 @@ public function validateRow($headers) {
       // The importer specified a strict requirement for number of columns in
       // the input file, but the header has more or less than that amount.
       return [
-        'case' => 'Headers provided does not have the expected number of headers',
+        'case' => self::$mapping['case-mismatch-count']['dev-case'],
         'valid' => FALSE,
         'failedItems' => $headers,
       ];
     }
 
     // At this point the headers input array is valid.
     return [
-      'case' => 'Headers exist and match expected headers',
+      'case' => self::$mapping['case-valid']['dev-case'],
       'valid' => TRUE,
       'failedItems' => [],
     ];
   }
 
+  /**
+   * Processes failed validation from ValidHeaders into a render array.
+   *
+   * @param array $validation_status
+   *   An associative array that was returned by the ValidHeaders validator in
+   *   the event of failed validation. It contains the following keys:
+   *   - 'case': a developer-focused string describing the case checked.
+   *   - 'valid': FALSE to indicate that validation failed.
+   *   - 'failedItems': an array of items that failed, either:
+   *     - 'headers': A string indicating the header row is empty.
+   *     - an array of column headers that was in the input file.
+   *       @see validateRow()
+   * @param array $metadata
+   *   An array of additional metadata (or contextual information) needed by the
+   *   process method. Here, the following keys are expected:
+   *   - 'column_headers': This contains an array of headers defined in the
+   *     importer. The index in this array MUST match the position
+   *     (starting with 0) of the column in the input file.
+   *     Eg: 'column_headers' => [
+   *           '2' => 'Header 2', // Header of column #3
+   *           '4' => 'Header 4', // Header of column #5
+   *         ];.
+   * @param array $tokens
+   *   [OPTIONAL] An array of values to use for token replacement.
+   *   @see ValidHeaders::$mapping
+   *   The following token keys will substitute the entire existing case message
+   *   to the user with the value of that token.
+   *   - 'case-empty-headers': the message when there is no header row.
+   *   - 'case-mismatch-values': the message when headers are not expected.
+   *   - 'case-mismatch-count': the message when headers count is not expected.
+   *   - 'num-expected-columns': number of expected column headers.
+   *
+   * @return array
+   *   A render array of type unordered list which is used to display feedback
+   *   to the user about the case that failed and the failed items from the
+   *   input file. This unordered list will include a table with a row of the
+   *   expected headers followed by a row of the provided headers.
```

Comment:
I think this can be improved a bit and hopefully is in line with what Lacey would suggest:
```suggestion
   *   A render array of type "unordered list" which is used to display feedback
   *   to the user about the validation failure. This process method renders a
   *   a single item consisting of a markup block with a message describing the
   *   case that was triggered, and a table with the following rows:
   *   - 'Expected Headers': the column header as it is expected by the importer
   *   - 'Provided Headers': the column header exactly as it was provided in the
   *     input file.
```